Readme.md
# Simple Logger

Lightweight, Simple library to write logs. Writes to console and file by default.

## Introduction

Simple logger is a simple library used to log to console and file. The usage is very simple, import the class and initialize it. You can use various logging methods available to create logs. Please note that initialization of logger is required and will throw an error when you use one of the available logging methods without first initializing the logs. If initialization was successful, it return `true` else throws an `error`. All operation related to files ex: writing to files, creation of file are synchronous.

```ts
import Logger from 'sds-simple-logger';

// Throw error if there was problem initializing logs using defaults
Logger.initLogs();

Logger.debug('Debug log');
Logger.info('Some info message');
Logger.warn('Some warning message');
Logger.error(new Error('Error message'));
```

Once imported and initialized, you can use it in any other file:

```ts
import Logger from 'sds-simple-logger';
// or
const Logger = require('sds-simple-logger');

Logger.log('Test logs');
```

Sample Output beautified JSON:

```ts
{
  "timestamp": "Fri, 15 Jan 2021 22:12:04 GMT",
  "type": "DEBUG",
  "message": "Fri Jan 15 2021 17:12:04 GMT-0500 (GMT-05:00) debug log"
}
```

String JSON output:

```ts
{"timestamp":"Fri, 15 Jan 2021 22:15:14 GMT","type":"LOG","message":"Test dir creation"}
```

String output:

```ts
Fri, 15 Jan 2021 22:14:17 GMT DEBUG: Fri Jan 15 2021 17:14:17 GMT-0500 (GMT-05:00) debug log
```

### Options:

You can pass various options as an options object when initializing the logger. Options available:

1. `cycleTime`: The time in milliseconds after which a new log file should be created, defaults to 86400000 ms or 24 hrs
2. `removeTime`: The time in milliseconds, log files older than this will be removed. It runs with the cycle time, defaults to 604800000 ms or 7 days
3. `logsDir`: The path and name of logs directory. Defaults to creation of **logs** directory in the current directory
4. `writeToFile`: By default all logs are written to file and console. If this is set to `false` logs will be only written to console. It is set to `true` by default.
5. `writeToConsole`: By default all logs are written to file and console. If this is set to false, logs will be only written to the logs file. It is set to `true` by default.
6. `json`: By default all logs are written as JSON object to console and file. You can set this to `false` to write log string instead.
7. `beautifyJson`: By default JSON object written to console and logs are beaufied or in other words formated for readibility. Is this option is set to `false`, logs will still be written in JSON format but will not beautified while writting.
8. `handleSyncErrors`: By default set to `true` logger while throw and error if a sync operation such as writting to file fails. If set to `false` the logger will not throw error but it will just print a error to the console.

Example:

```ts
Logger.initLogs({
  cycleTime: 86400000,
  removeTime: 86400000,
  logsDir: path.join(__dirname, '../test-logs'),
  logName: 'test.logs',
  writeToFile: true,
  writeToConsole: true,
  json: true,
  beautifyJson: true,
  handleSyncErrors: true,
});
```

### Methods Available:

All the methods available can take variable string messages or in other words are `variadic functions` except the error method, which takes only error object as input. For the error methods available you can pass a second argument i.e a boolean value of true. If you want to log the error and exit the Node process.

#### **Debug**

Prints **DEBUG** in front the logs. Sample output:
`Tue, 05 Jan 2021 06:12:49 GMT: DEBUG: Debug output test`

Usage:

```ts
// Signature of function
// (...messages: string[])
// Throw error if there was problem writing to logs

Logger.debug('Debug test');
Logger.debug('Debug test one', 'Debug test two', 'Debug test three');
```

#### **Info**

Prints **INFO** in front the logs. Sample output:
`Tue, 05 Jan 2021 06:13:18 GMT: INFO: Info output test`

Usage:

```ts
// Signature of function
// (...messages: string[])
// Throw error if there was problem writing to logs

Logger.info('Info test');
Logger.info('Info test one', 'Info test two', 'Info test three');
```

#### **Log**

Prints **LOG** in front the logs. Sample output:
`Tue, 05 Jan 2021 06:13:18 GMT: INFO: Info output test`

Usage:

```ts
// Signature of function
// (...messages: string[])
// Throw error if there was problem writing to logs

Logger.log('Log test');
Logger.log('Log test one', 'Log test two', 'Log test three');
```

#### **Warn**

Prints **WARN** in front the logs. Sample output:
`Tue, 05 Jan 2021 06:13:18 GMT: WARN: Warn output test`

Usage:

```ts
// Signature
// (...messages: string[])
// Throw error if there was problem writing to logs

Logger.warn('Warn test');
Logger.warn('Warn test one', 'Warn test two', 'Warn test three');
```

#### **Error**

Prints **Error** in front the logs and also prints the error stack. Along with this it also print a unique error key. If you want to exit the node process after printing the error pass `true` as the second arg. Sample output:

```
Tue, 05 Jan 2021 06:13:18 GMT: ERROR KEY: 1609827198551
Error: Error output test
  at main (/Users/someuser/Projects/Personal/Testing/index.ts:7:16)
  at Object.<anonymous> (/Users/someuser/Projects/Personal/Testing/index.ts:11:1)
  at Module.\_compile (internal/modules/cjs/loader.js:1015:30)
  at Module.m.\_compile (/usr/local/lib/node_modules/ts-node/src/index.ts:858:23)
  at Module.\_extensions..js (internal/modules/cjs/loader.js:1035:10)
  at Object.require.extensions.<computed> [as .ts] (/usr/local/lib/node_modules/ts-node/src/index.ts:861:12)
  at Module.load (internal/modules/cjs/loader.js:879:32)
  at Function.Module.\_load (internal/modules/cjs/loader.js:724:14)
  at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)
  at main (/usr/local/lib/node_modules/ts-node/src/bin.ts:227:14)
```

Usage:

```ts
// Signature
// (error: Error, exit: boolean = false): number;
// Throw error if there was problem writing to logs

Logger.error(new Error('Error test'));
// Exit the node process with non-zero exit code
Logger.error(new Error('Error test'), true);
```

#### **Prune Logs**

To prune logs i.e deleting the log directory and its content. You can use the `pruneLogs` method available on the logger instance. Method returns `true` if successfully deleted.

The next time you use a logging method available. The logs directory will be created according to the options passed in the `initLogs()` function.

```ts
Logger.pruneLogs();
```

### Express Request Logger Middelware

A small middleware function to log Express requests. Pass it before your routes to log all incoming requests. It will write to the same log file as the logger.

```ts
// Get the request logging middleware
const loggerMiddleware = Logger.expressReqLogs();

// Use the middleware with express app
app.use(loggerMiddleware);

// Output to console:
REQ LOGS: {"timestamp":"Tue, 12 Jan 2021 06:30:02 GMT","request":{"method":"GET","url":"test.com","clientIp":null,"ipFamily":null,"userAgent":null,"httpVersion":null,"params":{},"headers":{"token":"Test","random":"Random"},"body":{"username":"Test","password":"Test"}}}
```

#### Middleware options:

By default it pics the options passed in initLogs method of the logger, but those options can be overwritten by passing an options object. In other words, you can pass a options object when calling the `Logger.expressReqLogs()` to configure the logger middleware. Avaliable options are:

1. `writeToFile`: Boolean value, set to true by default
2. `writeToConsole`: Boolean value, set to true by default
3. `json`: Boolean value, set to true by default
4. `hideBodyFields`: Array of fields that you want to hide on the the request body. Hidden fields/keys and respective values won't be printed in logs
5. `hideHeaders`: Arrays of headers you want to hide

```ts
// Call with options
Logger.expressReqLogs({
  writeToFile: true,
  writeToConsole: true,
  json: true,
  hideBodyFields: ['password'],
  hideHeaders: ['token'],
  handleSyncErrors: true,
});

// Use the middleware with express app
app.use(loggerMiddleware);
```

Sample String output:

```ts
Fri, 15 Jan 2021 22:12:04 GMT, REQUEST: method - GET, url - test.com, clientIp - null, ipFamily - null, userAgent - null, httpVersion - null, params - {}, headers - {"random":"Random"}, body - {"username":"Test"}
```

Sample JSON string output:

```ts
{"timestamp":"Fri, 15 Jan 2021 22:15:14 GMT","type":"REQUEST","method":"GET","url":"test.com","clientIp":null,"ipFamily":null,"userAgent":null,"httpVersion":null,"params":{},"headers":{"random":"Random"},"body":{"username":"Test"}}
```

Sample Pretty JSON Output:

```ts
{
  "timestamp": "Fri, 15 Jan 2021 22:10:52 GMT",
  "type": "REQUEST",
  "method": "GET",
  "url": "test.com",
  "clientIp": null,
  "ipFamily": null,
  "userAgent": null,
  "httpVersion": null,
  "params": {},
  "headers": {
    "random": "Random"
  },
  "body": {
    "username": "Test"
  }
}
```


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v2.1.0](https://github.com/developerphilosophy/simple-logger/compare/v2.0.0...v2.1.0)

- MODIFIED: seggregated types, errors, and enums [`94fac37`](https://github.com/developerphilosophy/simple-logger/commit/94fac37628b86ffd6d34c931a5cb8c6cee803464)
- ADDED: Basic request logging middleware [`1179860`](https://github.com/developerphilosophy/simple-logger/commit/117986096b716de6e8c6641d2d8adeb42bbc87e4)
- ADDED: Json option and library function to write to console [`f865b0c`](https://github.com/developerphilosophy/simple-logger/commit/f865b0cc86e558a3653e25ee2952299b78add7d8)

### [v2.0.0](https://github.com/developerphilosophy/simple-logger/compare/v1.0.0...v2.0.0)

> 11 January 2021

- UPDATE: Simplified the working on logging functions and updated tests [`f68cd32`](https://github.com/developerphilosophy/simple-logger/commit/f68cd32e6e04069707bd071a68be61e8d011add5)
- ADDED: log method to the logger class and Readme documentation [`586f072`](https://github.com/developerphilosophy/simple-logger/commit/586f072d82d6fab565ce5ebcf117e2e4b9384419)
- ADDED: Additional test cases and simplified logic [`0231eeb`](https://github.com/developerphilosophy/simple-logger/commit/0231eebc3ff5783e8f505f3180ff252b80f5b2e5)

#### [v1.0.0](https://github.com/developerphilosophy/simple-logger/compare/v1.0.0-0...v1.0.0)

> 9 January 2021

- ADDED: Auto change log package to dev for recording changes logs [`896b6eb`](https://github.com/developerphilosophy/simple-logger/commit/896b6eb404f62b87b8f9b23befaf29c0a3765749)
- MODIFIED: Tests to add delay function [`72595c9`](https://github.com/developerphilosophy/simple-logger/commit/72595c9ac285c2f1e23840f8b9f8695a383190ee)
- FIX: Await added to async logging operations [`204444a`](https://github.com/developerphilosophy/simple-logger/commit/204444ad387f69de5314a4048da4121f91c77e6c)

#### [v1.0.0-0](https://github.com/developerphilosophy/simple-logger/compare/v0.0.9...v1.0.0-0)

> 9 January 2021

- BREAKING: Logging functions are no longer async [`1ea5498`](https://github.com/developerphilosophy/simple-logger/commit/1ea5498b57f579b36e40a317c81f481eec85b34e)
- BREAKING [`dbb376d`](https://github.com/developerphilosophy/simple-logger/commit/dbb376d3db026425744c24adc8b9ca0b077220b7)
- BREAKING: Premajor release, logging function no longer async [`105f4cd`](https://github.com/developerphilosophy/simple-logger/commit/105f4cd640eb396a3389a9f71f1eda56d6d5f9a8)

#### [v0.0.9](https://github.com/developerphilosophy/simple-logger/compare/v0.0.7...v0.0.9)

> 8 January 2021

- FIX: Merge conflicts [`255b50a`](https://github.com/developerphilosophy/simple-logger/commit/255b50a7bb773bf8c3f7a329ff49a564f5033c46)
- UPDATE: Package.json updated after moving dependencies to dev [`0fa6fb5`](https://github.com/developerphilosophy/simple-logger/commit/0fa6fb51562053f6d33a5add65cf4409d970f992)
- UPDATE: Moved dependencies to dev [`5bfbe16`](https://github.com/developerphilosophy/simple-logger/commit/5bfbe1630e276cd34af6bfca69326c4d9cd17433)

#### [v0.0.7](https://github.com/developerphilosophy/simple-logger/compare/v0.0.6...v0.0.7)

> 8 January 2021

#### [v0.0.6](https://github.com/developerphilosophy/simple-logger/compare/v0.0.4...v0.0.6)

> 6 January 2021

- UPDATE: Readme updated to add method example [`ecc89a6`](https://github.com/developerphilosophy/simple-logger/commit/ecc89a6df006119d9987b7ea19e45a96b60248cf)
- UPDATE: Readme format correction [`d94f863`](https://github.com/developerphilosophy/simple-logger/commit/d94f86373bbf8f95528a5699b585dbb8d1c2eeda)
- UPDATE: Minor changes to readme [`6770b62`](https://github.com/developerphilosophy/simple-logger/commit/6770b62a84c8fef4a54ec30c2fbac259a9ca81aa)

#### [v0.0.4](https://github.com/developerphilosophy/simple-logger/compare/v0.0.3...v0.0.4)

> 5 January 2021

- UPDATE: Patch version update to 4 [`0a5ff91`](https://github.com/developerphilosophy/simple-logger/commit/0a5ff917410cfd29b7b4d158d1bab2bec58fca84)

#### [v0.0.3](https://github.com/developerphilosophy/simple-logger/compare/v0.0.2...v0.0.3)

> 5 January 2021

- UPDATE: Patch version changed to 0.3 [`72a8e7e`](https://github.com/developerphilosophy/simple-logger/commit/72a8e7ea344d300477d68487ea548a72d632f97c)

#### [v0.0.2](https://github.com/developerphilosophy/simple-logger/compare/v0.0.1...v0.0.2)

> 5 January 2021

- ADD: Keywords to package.json [`ec7a74c`](https://github.com/developerphilosophy/simple-logger/commit/ec7a74c213ba46fc28c827aaa0c0653e11990a33)

#### v0.0.1

> 5 January 2021

- INIT: Initial setup for writing tests ready [`b883331`](https://github.com/developerphilosophy/simple-logger/commit/b883331709345c43c5c9b5ffa886609eee01f5ed)
- ADDED: Formatting and linting for code [`0539b6b`](https://github.com/developerphilosophy/simple-logger/commit/0539b6bec6a71b02d7464d340e437082ffd72bca)
- Intial commit [`6208c3c`](https://github.com/developerphilosophy/simple-logger/commit/6208c3c907db579249d152b964048793ac2893f6)


src/index.ts
import chalk from 'chalk';
import * as fs from 'fs';
import * as path from 'path';
import {
  printErrorToConsole,
  getUTCString,
  getLogObject,
  printJsonLog,
  printStringLog,
  logObjectToString,
  getErrorLogObject,
  printJsonError,
  printErrorString,
  printReqLogJson,
  printReqLogString,
  reqLogObjToString,
  errorObjectToString,
} from './lib/utilFunctions';
import { LogsNotInitaizedError, LogsAlreadyInitializedError, NotWritingAnyLogs } from './errors/errors';
import { Request, Response, NextFunction } from 'express';
import { LogType } from './enums/enums';
import { LogOptions, ReqMidOptions, ReqLogObject, LogObject } from './types/types';
import { clear } from 'console';

class SimpleLogger {
  private static _logName: string = 'sds-simple-logger.logs';
  private static _logsDir: string = path.join(process.env.PWD as string, 'logs'); // Deaults to current directory
  private static _isInitialized: boolean = false;
  private static _cycleTime: number = 86400000;
  private static _removeTime: number = 604800000;
  private static _writeToFile: boolean = true;
  private static _writeToConsole: boolean = true;
  private static _logsDirExists: boolean = false;
  private static _json: boolean = true;
  private static _beautifyJson: boolean = true;
  private static _handleSyncErrors: boolean = true;

  public static initLogs(opts: LogOptions = {}): boolean {
    if (this._isInitialized) throw LogsAlreadyInitializedError;
    /**
     * 1. Cycle time for when new log file is created
     * 2. Remove time, log files older than this are removed daily
     * 3. Name of the log file
     * 4. Path to log directory
     * 5. Should logs be written to file
     * 6. Should logs be writting to console
     * 7. Should output be in JSON format
     * 8. Do you want to handle sync errors yourself?
     */
    if (typeof opts.cycleTime !== 'undefined') this._cycleTime = opts.cycleTime;
    if (typeof opts.removeTime !== 'undefined') this._removeTime = opts.removeTime;
    if (typeof opts.logsDir !== 'undefined') this._logsDir = opts.logsDir;
    if (typeof opts.logName !== 'undefined') this._logName = opts.logName;
    if (typeof opts.writeToFile !== 'undefined') this._writeToFile = opts.writeToFile;
    if (typeof opts.writeToConsole !== 'undefined') this._writeToConsole = opts.writeToConsole;
    if (typeof opts.json !== 'undefined') this._json = opts.json;
    if (typeof opts.beautifyJson !== 'undefined') this._beautifyJson = opts.beautifyJson;
    if (typeof opts.handleSyncErrors !== 'undefined') this._handleSyncErrors = opts.handleSyncErrors;

    if (!this._writeToConsole && !this._writeToFile) {
      console.warn(`Both ${chalk.blue('writeToConsole')} and ${chalk.blue('writeToFile')} are set to false`);
      throw NotWritingAnyLogs;
    }

    try {
      if (!this._isInitialized) {
        this.createLogDir();
      }
      this.logFileInit();
      this.cycleLogs();
    } catch (error) {
      printErrorToConsole(error);
      throw error;
    }

    this._isInitialized = true;
    return true;
  }

  /**
   * Create the logs directory
   */
  private static createLogDir(): void {
    try {
      if (!fs.existsSync(this._logsDir)) {
        // If log directory does not exists, create one
        fs.mkdirSync(this._logsDir);
        this._logsDirExists = true;
      }
    } catch (error) {
      throw error;
    }
  }

  private static logFileInit(): void {
    const logName = path.join(this._logsDir, this._logName);
    try {
      fs.appendFileSync(logName, `\n*** New Logging Session Started Created on(UTC time): ${getUTCString()}***\n`);
    } catch (error) {
      throw error;
    }
  }

  private static cycleLogs(): void {
    try {
      this.logFileInit();
    } catch (error) {
      printErrorToConsole(error);
    }

    setInterval(() => {
      let date = new Date().toLocaleDateString();
      let time = new Date().toLocaleTimeString();
      time = time.replace(/ /g, '').replace(/:/g, '');
      date = date.replace(/\s+|[,\/]/g, '');

      const oldPath = path.join(this._logsDir, `${this._logName}`);
      const newPath = path.join(this._logsDir, `${this._logName}.${date}.${time}`);

      fs.rename(oldPath, newPath, (error) => {
        if (error) printErrorToConsole(error);
        this.logFileInit();
      });
      this.removeOldLogs();
    }, this._cycleTime);
    // For 24 hours: 86400000
  }

  private static removeOldLogs(): void {
    fs.readdir(this._logsDir, (errorOne, files) => {
      if (errorOne) printErrorToConsole(errorOne);
      files.forEach((file, index) => {
        fs.stat(path.join(this._logsDir, file), (errorTwo, stat) => {
          if (errorTwo) return printErrorToConsole(errorTwo);
          const now = new Date().getTime();
          const endTime = new Date(stat.ctime).getTime() + this._removeTime;
          // For 7 days: 604800000
          if (now > endTime) {
            fs.unlink(path.join(this._logsDir, file), (errorThree) => {
              if (errorThree) return printErrorToConsole(errorThree);
              const dateTime = new Date().toLocaleString();
              return console.log(chalk.green(`${dateTime}: Olds logs pruned successfully`));
            });
          }
        });
      });
    });
  }

  private static writeToLogFile(message: any, json: boolean = true): void {
    const fullLogPath: string = this.getFullLogPath();

    try {
      if (!this._logsDirExists || !fs.existsSync(this._logsDir)) {
        fs.mkdirSync(this._logsDir, { recursive: true });
      }
      if (json) {
        const jsonString = this._beautifyJson
          ? `${JSON.stringify(message, null, 2)},\n`
          : `${JSON.stringify(message)},\n`;
        fs.appendFileSync(fullLogPath, jsonString);
      } else {
        fs.appendFileSync(fullLogPath, `${message}\n`);
      }
    } catch (error) {
      throw error;
    }
  }

  private static getFullLogPath(): string {
    return path.join(this._logsDir, this._logName);
  }

  private static writeLogsFactory(logObj: LogObject) {
    if (this._writeToConsole) {
      if (this._json) {
        printJsonLog(logObj);
      } else {
        printStringLog(logObj);
      }
    }

    if (this._writeToFile) {
      if (this._json) {
        this.writeToLogFile(logObj, true);
      } else {
        const logString = logObjectToString(logObj);
        this.writeToLogFile(logString, false);
      }
    }
  }

  /**
   * Exposed functions
   */
  public static pruneLogs(): boolean {
    if (!this._isInitialized) throw LogsNotInitaizedError;

    try {
      // 1. Delete the content of directory
      const dirContent = fs.readdirSync(this._logsDir);
      dirContent.forEach((file) => {
        fs.unlinkSync(path.join(this._logsDir, file));
      });
      // 2. Remove the directory
      fs.rmdirSync(this._logsDir);
      this._logsDirExists = false;
      return true;
    } catch (error) {
      if (this._handleSyncErrors) {
        throw error;
      } else {
        printErrorToConsole(error);
        return false;
      }
    }
  }
  /**
   * All the logging functions
   */
  public static debug(...messages: string[]): void {
    if (!this._isInitialized) throw LogsNotInitaizedError;

    try {
      for (const message of messages) {
        const logObj = getLogObject(LogType.DEBUG, message);
        this.writeLogsFactory(logObj);
      }
    } catch (error) {
      if (this._handleSyncErrors) {
        throw error;
      } else {
        printErrorToConsole(error);
      }
    }
  }

  public static info(...messages: string[]): void {
    if (!this._isInitialized) throw LogsNotInitaizedError;
    try {
      for (const message of messages) {
        const logObj = getLogObject(LogType.INFO, message);
        this.writeLogsFactory(logObj);
      }
    } catch (error) {
      if (this._handleSyncErrors) {
        throw error;
      } else {
        printErrorToConsole(error);
      }
    }
  }
  public static log(...messages: string[]): void {
    if (!this._isInitialized) throw LogsNotInitaizedError;
    try {
      for (const message of messages) {
        const logObj = getLogObject(LogType.LOG, message);
        this.writeLogsFactory(logObj);
      }
    } catch (error) {
      if (this._handleSyncErrors) {
        throw error;
      } else {
        printErrorToConsole(error);
      }
    }
  }

  public static warn(...messages: string[]): void {
    if (!this._isInitialized) throw LogsNotInitaizedError;
    try {
      for (const message of messages) {
        const logObj = getLogObject(LogType.WARN, message);
        this.writeLogsFactory(logObj);
      }
    } catch (error) {
      if (this._handleSyncErrors) {
        throw error;
      } else {
        printErrorToConsole(error);
      }
    }
  }

  public static error(error: Error, exit: boolean = false): number {
    if (!this._isInitialized) throw LogsNotInitaizedError;

    let errorKey = 0;

    try {
      const errorLogObject = getErrorLogObject(error);
      errorKey = errorLogObject.errorKey;

      if (this._writeToConsole) {
        if (this._json) {
          printJsonError(errorLogObject);
        } else {
          printErrorString(errorLogObject);
        }
      }

      if (this._writeToFile) {
        if (this._json) {
          this.writeToLogFile(errorLogObject, true);
        } else {
          const errorString = errorObjectToString(errorLogObject);
          this.writeToLogFile(errorString, false);
        }
      }
      return errorLogObject.errorKey;
    } catch (error) {
      if (this._handleSyncErrors) {
        throw error;
      } else {
        printErrorToConsole(error);
        return errorKey;
      }
    }
  }

  public static expressReqLogs(opts: ReqMidOptions = {}) {
    return (req: Request, res: Response, nxt: NextFunction) => {
      if (!this._isInitialized) throw LogsNotInitaizedError;
      const { headers, httpVersion, method, socket, url, body, params } = req;
      const { remoteAddress, remoteFamily } = socket;

      const requestMethod = method || null;
      const requestUrl = url || null;
      const requestIpAddress = remoteAddress || null;
      const requestUserAgent = req.get('user-agent') || null;
      const requestBody = req.body || null;
      const requestParams = req.params || null;
      const requestHeaders = headers || null;
      const requestHttpVersion = httpVersion || null;
      const requestRemoteFamily = remoteFamily || null;

      let writeToFile: boolean = this._writeToFile;
      let writeToConsole: boolean = this._writeToConsole;
      let json: boolean = this._json;
      let hideBodyFields: string[] = [];
      let hideHeaders: string[] = [];
      let handleSyncErrors = this._handleSyncErrors;

      if (typeof opts.writeToFile !== 'undefined') writeToFile = opts.writeToFile;
      if (typeof opts.writeToConsole !== 'undefined') writeToConsole = opts.writeToConsole;
      if (typeof opts.hideBodyFields !== 'undefined') hideBodyFields = opts.hideBodyFields;
      if (typeof opts.hideHeaders !== 'undefined') hideHeaders = opts.hideHeaders;
      if (typeof opts.json !== 'undefined') json = opts.json;
      if (typeof opts.handleSyncErrors !== 'undefined') handleSyncErrors = opts.handleSyncErrors;

      /**
       * Delete hiddens fields based on middleware settings
       */
      if (hideBodyFields.length > 0) {
        hideBodyFields.forEach((field) => {
          delete body[field];
        });
      }

      if (hideHeaders.length > 0) {
        hideHeaders.forEach((field) => {
          delete requestHeaders[field];
        });
      }

      const dataObject: ReqLogObject = {
        timestamp: getUTCString(),
        type: LogType.REQUEST,
        method: requestMethod,
        url: requestUrl,
        clientIp: requestIpAddress,
        ipFamily: requestRemoteFamily,
        userAgent: requestUserAgent,
        httpVersion: requestHttpVersion,
        params: requestParams,
        headers: requestHeaders,
        body: requestBody,
      };

      try {
        if (writeToConsole) {
          if (json) {
            printReqLogJson(dataObject);
          } else {
            printReqLogString(dataObject);
          }
        }

        if (writeToFile) {
          if (json) {
            this.writeToLogFile(dataObject, true);
          } else {
            const errorString = reqLogObjToString(dataObject);
            this.writeToLogFile(errorString, false);
          }
        }

        return nxt();
      } catch (error) {
        if (handleSyncErrors) {
          return nxt(error);
        } else {
          printErrorToConsole(error);
          return nxt();
        }
      }
    };
  }
}

export default SimpleLogger;
